### ISSUE TITLE
Add catalysts.sbt project file

### ISSUE THREAD

**Issue by InTheNow** [InTheNow/sbt-catalysts#33]
Add the required project details to a dedicated file. Eg

```scala
catalystsProject in ThisBuild := CatalystsGitProject(
  project      = "catalysts",
  properName   = "Catalysts",
  organisation = "InTheNow",
  publishedTo  = "org.typelevel",
  licenses     = List(apache),
  authors      = List(developer(name = "Alistair Johnson", id = "inthenow"))
)
```

@milessabin @dwijnand Does this look OK, please?

**Comment by dwijnand** [comment]
Depends on what problem are you going to solve with this.

I would only introduce alternatives to how sbt does things currently if there is some massive improvement.

As always I present sbt-typelevel as evidence for trying to abstract too far away from regular sbt.

**Comment by milessabin** [comment]
I think this looks very promising. :+1:

**Comment by InTheNow** [comment]
@dwijnand @milessabin 

Thanks for feedback, here and elsewhere. 

Summing up the comments, I have the feeling that we are actually converging on a general path forward, namely providing some common patterns to facilitate Typelevel builds without moving away from pure SBT. And the two edges of that pattern are very well represented by your two opinions, please keep them coming ;)

I think the easiest route forward is for me to implement the requirements we have as it is not really clear what the individual little bits do, how they help and how they naturally fit with pure SBT. If we aim to stay close to SBT, it should not be a lot of work.... 

Just taking the example above, all it does is populate standard SBT settings and generates values for a few more. But the main reason to have this done via a method/constructor is just to force the settings to be made. Perhaps "force" is too strong, maybe "self documenting the minimum SBT settings you need to set, out of all the many available ones"

**Comment by dwijnand** [comment]
I agree with you, I think we are converging. I think we should strive to remain modular, and not provide a "our playground, our rules" / framework-like feel.

As a simple, on-the-spot example, given one wants to use sbt-catalysts right at the start of a project (because of an assortment of benefits it brings over base sbt), they might find it a little off-putting to have to deciding right upfront on licenses, authors, publish repo, etc. They ultimately need to, but I'm just thinking with my UX hat on: are they happy to provide those on project construction.

To that end I was thinking: what about if sbt-catalysts provided a "doctor"/"lint" like task/command, which would provide info/warnings for these things, with helpful code snippets. This is in part pulling in an idea I've had on my TODO list for a long time, a linter-like tool for sbt builds.

Note, I might be completely wrong, perhaps most library authors are happy to get this all out the way at the start. Given your experience over several projects, what's your experience been?

**Comment by milessabin** [comment]
I think @InTheNow's characterization of us as at opposite ends of the sbt spectrum is about right: I've described my process of constructing builds as "cargo culting boilerplate" ;-) I really do think there's value in capturing that boilerplate as higher level functionality that can be reused across Typelevel projects, and I think there are plenty of other people in my predicament who would agree. I take your point that sbt-typelevel didn't succeed, but think it's hasty to generalize from that one data point.

On your specific points, I think I agree that `publishedTo` shouldn't be mandatory up front. But I do think that it's reasonable to insist on a license and an author. I agree that this is a little prescriptive, but I think that's part of what sbt-catalysts is about.

As far as the license is concerned I think it would make sense to document a couple of recommendations (I would suggest recommending Apache 2 for BSD-style and (L)GPL for copyleft).

I do think a linter is a good idea, but I'm not persuaded that it would be desirable to have it produce code snippets to be copy and pasted into a build.sbt. I think that that would encourage cargo cult copy/paste and make for unmaintainable builds.

**Comment by InTheNow** [comment]
I like the idea of having one object that contains all the mandatory info, including `publishedTo`. Of course, it could be empty and checked later - not encouraged but.... Or defaulted to `org.typelevel` (and license defaulted too)

More likely for a new project, we would have some tool that produces this file based on git and filesystem info. Migrating projects would likely have this file fully populated straight away.

**Comment by milessabin** [comment]
On `publishedTo`, I think what you say makes sense for established projects, what about the first cut at a new project where there's no immediate idea of publishing at all, just a trail balloon? Or maybe a project associated with talk or a training course where publishing isn't appropriate?

**Comment by InTheNow** [comment]
@milessabin fair point. I've been thinking some more about this in more detail. I think the issue here is my naming, and the mapping between Github concepts and sbt ones. I'll spec this out more clearly, but for now, `organisation` here is the GiHub Organisation, `publishedTo` is the sbt `organisation`

In an sbt project use `projectInfo` or `show projectInfo`, [from] (https://github.com/sbt/sbt/blob/e79f1b3a159fd2fcb2eabd980ff56c923300d22f/main/src/main/scala/sbt/Defaults.scala#L1181)

It would be cool to get a nice way to set things up for users to set this all up in a nice way, and really that is just the goal here.

**Comment by milessabin** [comment]
Could there be an interactive task which prompts people for the basics and writes an initial catalysts.sbt? I'm thinking of interactivity in the sense of sbt-release rather than anything fancier.

**Comment by dwijnand** [comment]
There's some cross-over for that with (fairly new) https://github.com/sbt/sbt-fresh.

TIL about `projectInfo`.

**Comment by milessabin** [comment]
Yes, definitely! Thanks for the pointer :-)

**Comment by InTheNow** [comment]
Absolutely, that's actually what I meant ^^ "some tool", and yes, had sbt-fresh in mind

**Comment by InTheNow** [comment]
Ref https://github.com/ensime/ensime-server/issues/1222

Ensime's module have different licenses, so the mock details ^^ would not suffice as that is `in ThisBuild` only. Settings should be per module, as SBT.
